### PR TITLE
Wmf

### PR DESCRIPTION
**Pull request by mbeckert**
### host-app

**webpack/base.js**
```
ModuleFederationPlugin = require("webpack/lib/container/ModuleFederationPlugin");
[...]
new ModuleFederationPlugin({
  name: 'host',
  remotes: {},
  shared: Object.keys(packageJson.dependencies),
}),
```
**src/routes/Home.tsx**
`<SwidgetLoader url="http://127.0.0.1:7070/swidgetEntry.js" scope="swidget" module="app" />`

### swidget-app

**webpack/base.js**
```
ModuleFederationPlugin = require("webpack/lib/container/ModuleFederationPlugin");
[...]
new ModuleFederationPlugin({
  name: 'swidget',
  filename: "swidgetEntry.js",
  exposes: {
    'app': './src/App',
  },
  shared: Object.keys(packageJson.dependencies),
}),
```

## router
* integrated universal-router
* see https://github.com/kriasoft/universal-router/blob/main/docs/api.md
* implemented 3 URL-serialization-strategies (path, hash, query)

### route-declaration
**NEW**
```
container.bind<RouteConfigType>(serviceIds.routes).toConstantValue([
    {
      name: 'home',
      path: '',
      action: () => {
        return () => <h1>Hello World</h1>
      },
    }
  ]);
```

**OLD**
```
  container.bind<IRouteConfig[]>(serviceIds.routes).toConstantValue([
    {
      component: () => <h1>Hello World</h1>,
      name: 'home',
      pattern: '/',
    },
  ]);
```
### service
* added ```navigateTo()```
* marked ```linkToHome()``` and ```navigateToHome()``` as deprecated since there is no fix home-route
### route-serializers
**Path (default)**
* http://127.0.0.1:8080/demoContent
```
container.bind(serviceIds.routerSerializeRouteInUrlStrategy).toConstantValue(
  new Path()
);
```
**QueryString (default as swidget)**
* http://127.0.0.1:8080/?example=%2FdemoContent
```
container.bind(serviceIds.routerSerializeRouteInUrlStrategy).toConstantValue(
  new QueryString('example')
);
```
**Hashbang**
* http://127.0.0.1:8080/#!%2FdemoContent
```
container.bind(serviceIds.routerSerializeRouteInUrlStrategy).toConstantValue(
  new Hashbang()
);
```
### Link-Component
* renders anchor-element
```
<Link name="demoContent" query={{ gordon: 'shumway' }} parameter={{ le: 'go' }}>
  link
</Link>
```

### CODE REVIEW

File: boilerplate/.build/webpack/base.js [hunk 1]
```diff
@@ -25,10 +21,17 @@ const base = {
...
   output: {
     path: path.resolve(__dirname, '../../dist/app'),
-    filename: devMode ? 'swidget.js' : `${packageName}_${version}_[name].js`,
+    filename: devMode ? 'app.js' : `${packageName}_${version}_[name].js`,
   },
   module: {
     rules: [
+      {
+        test: /bootstrap\.tsx$/,
+        loader: 'bundle-loader',
```
Comment (BlackDark) [inline boilerplate/.build/webpack/base.js:30]: für was ist das?

File: boilerplate/.build/webpack/prod.js [hunk 2]
```diff
@@ -22,9 +22,6 @@ const prodConfig = {
   name: 'ProdConfig',
   mode: 'production',
   devtool: sourceMapGeneration,
-  output: {
-    publicPath: packageJson.config.publicPath || '',
```
Comment (BlackDark) [inline boilerplate/.build/webpack/prod.js:26]: war das nicht notwendig um auf subPath deployen zu können?
wenn ich beispiele unter `https//host/meineApp` das bereiten möchte

**Comment by mbeckert** [comment]
|                  test-scenario                   | DEV classic hotloading | PROD classic hotloading | DEV webpack module federation | PROD webpack module federation |
|--------------------------------------------------|------------------------|-------------------------|-------------------------------|--------------------------------|
| nested routing                                   | (x)                    | (x)                     | (x)                           | (x)                            |
| swidget has further dependency (unknown to host) | (x)                    | (x)                     | (x)                           | (x)                            |
| swidget consumes DI Binding from host            | (x)                    | (x)                     | (x)                           | (x)                            |
| swidget re-binds DI Binding from host            | (x)                    | (x)                     | (x)                           | (x)                            |
| interconnection                                  | (x)                    | (x)                     | (x)                           | (x)                            |

File: core/src/core/registerDefaultDependencies.tsx [hunk 3]
```diff
@@ -20,19 +20,19 @@ function registerDefaultDependencies(container: IDiContainer, name: string) {
...
+        return () => <h1>Hello World</h1>
+      },
+    }
   ]);
 
-  container.bind(serviceIds.routerUrlFormatStrategy).toDynamicValue(() => {
-    return new DefaultRouterStrategy(null, null);
-  });
+  container.bind(serviceIds.routerSerializeRouteInUrlStrategy).toConstantValue(
+    new Path()
```
Comment (BlackDark) [inline core/src/core/registerDefaultDependencies.tsx:34]: PathSerializer ist ja der "Standard" jetzt. Hast du mal getestet, wie es ist, wenn man in irgendeiner route ist und dann mit F5 neu lädt?
Soweit ich weiß funktioniert das mit SPAs nicht, da wir folderstruktur nicht abbilden sondern das alles code seitig gerouted werden muss.

File: core/src/router/component/RouterProvider.tsx [hunk 4]
```diff
@@ -2,13 +2,109 @@
...
+    });
+  }, [])
+
+  if (!activeRoute) {
+    return null;
+  }
+
+  const Route = activeRoute.component;
+
+  return (
```
Comment (BlackDark) [inline core/src/router/component/RouterProvider.tsx:69]: Weißt du, ob das mit SSR (ServiceSideRendering) funktionieren würde?

**Comment by BlackDark** [comment]
@mbeckert rebase please to resolve conflicts

**Comment by prode81** [comment]
First findings:

- When adding a swidget, we get following error in console on route-changes:
`Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
       in RouterProvider (created by swidget)`
- Swidget crashes after a few route-changes.
- Router replaces / to %2F, when using query or hashbang strategy (not an error, but if possible, this could be optimized)

File: packages/boilerplate-host/.build/webpack/prod.js [hunk 5]
```diff
@@ -42,14 +40,6 @@ const prodConfig = {
     new SimpleProgressWebpackPlugin({
       format: process.env.verbose === 'true' ? 'expanded' : 'compact',
     }),
-    new faviconsWebpackPlugin({
```
Comment (BlackDark) [inline packages/boilerplate-host/.build/webpack/prod.js:48]: Hatten wir das nicht für das Icon gebraucht?

**Review by BlackDark — APPROVED** [review]

**Pull request merged**